### DISCUSSION TITLE
SMTP sunset

### DISCUSSION THREAD

**Discussion by amasterengineer** [dani-garcia/vaultwarden#5084] [Ideas]
Many email providers are sunsetting basic authentication to include smtp. Will this application be updated to allow email notifications via moderen authentication, that outlook and Gmail will be supporting after fall 2024.

**Comment by BlackDex** [comment]
See https://github.com/dani-garcia/vaultwarden/discussions/4518#discussioncomment-9211666

**Reply by BlackDex** [reply]
Adding something like this takes a some effort. 
Also, using a proxy like this could be beneficial if using more applications and then only needing to configure one item extra.

**Reply by Norsagir** [reply]
PS : As we talked earlier BlackDex, I tried your "email-oauth2-proxy" solution. It does not work.
I opened a discussion on their github, here is the answer : 

> Personal (i.e., free) accounts work, but you'll need to use an existing OAuth 2.0 client ID that is already admin approved by Microsoft as you can't add any yourself. See the proxy readme for links to existing open source clients you could adopt.

And here is their readme : 
> If you have an existing client ID and secret for a desktop app, you can use these directly in the proxy. If this is not possible, you can also reuse the client ID and secret from any email client that supports IMAP/POP/SMTP OAuth 2.0 authentication with the email server you would like to connect to (such as [the](https://github.com/mozilla/releases-comm-central/blob/812b7c9068ca5cac0580b0ddbea8e34c141cd441/mailnews/base/src/OAuth2Providers.jsm) [many](https://github.com/mozilla/releases-comm-central/blob/master/mailnews/base/src/OAuth2Providers.sys.mjs) [existing](https://github.com/Foundry376/Mailspring/blob/master/app/internal_packages/onboarding/lib/onboarding-constants.ts) [open](https://gitlab.gnome.org/GNOME/evolution-data-server/-/blob/master/CMakeLists.txt) [source](https://gitlab.gnome.org/GNOME/gnome-online-accounts/-/blob/master/meson_options.txt) [clients](https://github.com/M66B/FairEmail/blob/master/app/src/main/res/xml/providers.xml) with OAuth 2.0 support), but please do this with care and restraint as access through reused tokens will be associated with the token owner rather than your own client.

So, for my case, I have to use an external client to use email-oauth2-proxy to use Vaultwarden.

I know implementing this takes a lot of efforts, but I think that it will become necessary sooner or later.
For now I created another email address thanks to a small email provider who has not banned basic authentication.
But since it's the right way of upgrading the email solution, as it is way more secure, I think Vaultwarden has to be compatible with OAuth2.

Again, I'm not a dev, so I'm really sorry I can't help you implementing this, but please take all your users' discussions about SMTP into account.

**Reply by BlackDex** [reply]
Well, in general we accept well written PR's and it's free for users to add this.
Also there are other solutions not using such email providers which still support SMTP.

And, it might be even better if this is going to be added to `lettre` instead of only Vaultwarden so that everybody who uses the `lettre` crate can use this.

**Reply by Norsagir** [reply]
Sorry but I think that you don't read users comments or maybe you don't want to discuss this point.

As I said concerning the PRs (I see that you use this sentence a lot), :
> Again, I'm not a dev, so I'm really sorry I can't help you implementing this

Also, telling people there are other solutions not using some email providers is a wrong idea, because I might tell that there are other passwords vaults as well.
I'm a big fan of Vaultwarden, if the "there are other solutions" point is completely true, I think it's the wrong way of developping a project.

But, to end this discussion as I think you don't really want to implement OAuth2, I agree with your last sentence.
As Vaultwarden is using `lettre`, maybe it's a good option to ask them how the OAuth2 implementation goes.

**Reply by BlackDex** [reply]
I do read the comments, and i do understand that you are not a dev. That doesn't mean someone else might be able to do this.

Also, this project is done during the spare time people have. It's not that this is our daily job.
Stating that you (or other people) are not a developer and need this doesn't make me jump and build it right away, if that was the case i would be working on something else almost every day without finishing anything.

There are other features also needed for Vaultwarden, like the new manger roles for example, which is also still not done and a needed feature for Vaultwarden to be able to use newer web-vault versions which have bug fixes and newer features for example.

And you assuming **I** don't want to have this implemented is a bit offending.
If i had plenty of time i would try to implement it a.s.a.p. And i would even do that for Lettre if that is the right place for it.
As i do not have plenty of time (since i have full time job and other activities) i need to select items which i deem important to work on first, and sometimes even items which i might enjoy to work on, as it must stay fun for the most part of the time.

To help in the meantime i try to search for solutions so people are able to keep using there SMTP provider. If that means you need to extract a client id/secret from somewhere to make that work, and you do not want to do that, that is your choice. And i provide other solutions to try and work around that. If that also does not suite your needs, then I'm afraid we have to wait until either my self, or someone else starts working on implementing this in either Vaultwarden or Lettre.

Also, this feature request is already added to #246 which has a list of other items too.

**Reply by paolobarbolini** [reply]
EDIT: I didn't see the reply until I posted this.

I think what you're saying is clear, but it's important to point out that for most of us, this is just a hobby. While we may like and agree with the ideas being suggested (it'd be cool to have OAuth2 in lettre but I currently don't have the time either to implement it), we are under no obligation (and simply **may not have the time or interest**) to implement them.

The options are:

1. Contribute directly - Develop and submit the changes yourself. This won't work if you are not a developer, but it's not the only option.
2. Spread awareness - Grow interest in the community and have someone else contribute the change.
3. Sponsor the development - Sponsor one of the maintainers or someone else to implement it
4. Do nothing - Hope the problem goes away.
5. Apply a workaround - Like using "email-oauth2-proxy", or a different provider.
6. Use another project - Accepting a different set of advantages and disadvantages.

Please understand that like any business or endeavor in general works, as maintainers we have to make trade-offs because it's usually not feasible to do everything the community asks for, or even what's on our minds sometimes.

**Reply by Norsagir** [reply]
I do understand what you two are saying, I know most of github things are hobbies.
But I pointed out the fact that you need to understand users too :)
For example, this message : https://github.com/dani-garcia/vaultwarden/discussions/4518#discussioncomment-10625042
I'm in that precise case, but all I did was deploying Vaultwarden and adding an entry in the basic Synology reverse proxy. Many users do not want to add multiple solutions just to fix an email problem...

Peace, I'm friendly and I want to make Vaultwarden better ! I just wanted to point out the fact that when users take their time to write something on github, please be patient, as many of us are not as skillful as you guys !

**Comment by amasterengineer** [comment]
I am sorry for what seems to be adding conflict, just wanted to bring this up as I just learned about it. I am more then fine with just using a Gmail account for my smtp needs if they will keep supporting it. Currently gmail is working. love Vaultwarden and just wanted to bring some light to an issue that I didn't expect to be an issue as I didn't know smtp was under fire by some of the email providers.
